File: core/src/main/java/com/metricstream/jdbc/JdbcSQLBuilderProvider.java
/*
 * Copyright © 2020-2021, MetricStream, Inc. All rights reserved.
 */
package com.metricstream.jdbc;

import java.sql.Timestamp;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import java.math.BigDecimal;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;


final class JdbcSQLBuilderProvider implements SQLBuilderProvider {

    private PreparedStatement build(SQLBuilder sqlBuilder, Connection connection) throws SQLException {
        return build(sqlBuilder, connection, (String[]) null);
    }

    private PreparedStatement build(SQLBuilder sqlBuilder, Connection connection, String... columns) throws SQLException {
        ArrayList<Object> expanded = null;
        if (!sqlBuilder.arguments.isEmpty()) {
            expanded = new ArrayList<>(sqlBuilder.arguments.size());
            int sqlPos = 0;
            for (Object arg : sqlBuilder.arguments) {
                sqlPos = sqlBuilder.statement.indexOf("?", sqlPos) + 1;
                if (sqlPos == 0) {
                    // We ran out of placeholders (i.e. we have extra parameters).
                    // We do not consider that as a bug (though one could argue this
                    // should result in a warning).
                    break;
                }
                if (arg instanceof Collection) {
                    @SuppressWarnings("unchecked")
                    final Collection<Object> col = (Collection<Object>) arg;
                    final int length = col.size();
                    if (length == 0) {
                        throw new SQLException("Collection parameters must contain at least one element");
                    }
                    // The statement already contains one "?", therefore we start with 1 instead of 0
                    for (int k = 1; k < length; k++) {
                        sqlBuilder.statement.insert(sqlPos, ",?");
                    }
                    // move sqlPos beyond the inserted ",?"
                    sqlPos += 2 * (length - 1);
                    expanded.addAll(col);
                } else {
                    expanded.add(arg);
                }
            }
        }

        sqlBuilder.interpolate(true);
        final PreparedStatement ps;
        if (columns == null || columns.length == 0) {
            ps = connection.prepareStatement(sqlBuilder.statement.toString(), sqlBuilder.resultSetType, ResultSet.CONCUR_READ_ONLY);
        } else {
            ps = connection.prepareStatement(sqlBuilder.statement.toString(), columns);
        }
        try {
            if (sqlBuilder.fetchSize > 0) {
                ps.setFetchSize(sqlBuilder.fetchSize);
            }
            if (sqlBuilder.maxRows >= 0) {
                ps.setMaxRows(sqlBuilder.maxRows);
            }

            if (expanded != null && !expanded.isEmpty()) {
                int idx = 0;
                for (Object arg : expanded) {
                    if (arg instanceof LongString) {
                        ps.setCharacterStream(++idx, ((LongString) arg).getReader());
                    } else if (arg instanceof SQLBuilder.Masked) {
                        ps.setObject(++idx, ((SQLBuilder.Masked) arg).data);
                    } else {
                        ps.setObject(++idx, arg);
                    }
                }
            }
        } catch (SQLException ex) {
            SQLBuilder.close(ps);
            throw ex;
        }
        return ps;
    }

    @Override
    public ResultSet getResultSet(SQLBuilder sqlBuilder, Connection connection, boolean wrapConnection) throws SQLException {
        PreparedStatement preparedStatement = null;
        try {
            preparedStatement = build(sqlBuilder, connection);
            return wrapConnection
                   ? SQLBuilder.wrapConnection(preparedStatement.executeQuery())
                   : SQLBuilder.wrapStatement(preparedStatement.executeQuery());
        } catch (SQLException e) {
            SQLBuilder.close(preparedStatement);
            throw e;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public int getInt(SQLBuilder sqlBuilder, Connection connection, int columnNumber, int defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getInt(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public int getInt(SQLBuilder sqlBuilder, Connection connection, String columnName, int defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getInt(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public long getLong(SQLBuilder sqlBuilder, Connection connection, int columnNumber, long defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getLong(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public long getLong(SQLBuilder sqlBuilder, Connection connection, String columnName, long defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getLong(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public double getDouble(SQLBuilder sqlBuilder, Connection connection, int columnNumber, double defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getDouble(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public double getDouble(SQLBuilder sqlBuilder, Connection connection, String columnName, double defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getDouble(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public String getString(SQLBuilder sqlBuilder, Connection connection, int columnNumber, String defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getString(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public BigDecimal getBigDecimal(SQLBuilder sqlBuilder, Connection connection, String columnName, BigDecimal defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getBigDecimal(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public BigDecimal getBigDecimal(SQLBuilder sqlBuilder, Connection connection, int columnNumber, BigDecimal defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getBigDecimal(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public String getString(SQLBuilder sqlBuilder, Connection connection, String columnName, String defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getString(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Object getObject(SQLBuilder sqlBuilder, Connection connection, int columnNumber, Object defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnNumber) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Object getObject(SQLBuilder sqlBuilder, Connection connection, String columnName, Object defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public OffsetDateTime getDateTime(SQLBuilder sqlBuilder, Connection connection, int columnNumber, OffsetDateTime defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnNumber, OffsetDateTime.class) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public OffsetDateTime getDateTime(SQLBuilder sqlBuilder, Connection connection, String columnName, OffsetDateTime defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnName, OffsetDateTime.class) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Instant getInstant(SQLBuilder sqlBuilder, Connection connection, int columnNumber, Instant defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnNumber, OffsetDateTime.class).toInstant() : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Instant getInstant(SQLBuilder sqlBuilder, Connection connection, String columnName, Instant defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getObject(columnName, OffsetDateTime.class).toInstant() : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Timestamp getTimestamp(SQLBuilder sqlBuilder, Connection connection, String columnName, Timestamp defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getTimestamp(columnName) : defaultValue;
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Override
    public Timestamp getTimestamp(SQLBuilder sqlBuilder, Connection connection, int columnNumber, Timestamp defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rs.getTimestamp(columnNumber) : defaultValue;
        }
    }

    /**
     * Executes the SQL statement.
     * @param connection The Connection from which the PreparedStatement is created
     * @return The result of executeUpdate of that statement
     * @throws SQLException the exception thrown when executing the query
     */
    @Override
    public int execute(SQLBuilder sqlBuilder, Connection connection) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection)) {
            return ps.executeUpdate();
        }
    }

    /**
     * Executes the SQL statement.
     * @param connection The Connection from which the PreparedStatement is created
     * @param keyColumns column names from the underlying table for which the inserted values will be returned.  Note that these names
     *               not necessarily have to be part of the keyColumns into which the builder explicitly inserts values.
     * @return The result of getGeneratedKeys of that statement
     * @throws SQLException the exception thrown when executing the query
     */
    @Override
    public ResultSet execute(SQLBuilder sqlBuilder, Connection connection, String... keyColumns) throws SQLException {
        PreparedStatement ps = build(sqlBuilder, connection, keyColumns);
        ps.executeUpdate();
        return SQLBuilder.wrapStatement(ps.getGeneratedKeys());
    }

    @Override
    public <T> List<T> getList(SQLBuilder sqlBuilder, Connection connection, SQLBuilder.RowMapper<T> rowMapper, boolean withNull) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return getList(rs, rowMapper, withNull);
        }
    }

    @Override
    public <K, V> Map<K, V> getMap(SQLBuilder sqlBuilder, Connection connection,
            SQLBuilder.RowMapper<Map.Entry<K, V>> rowMapper, boolean withNull) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return getMap(rs, rowMapper, withNull);
        }
    }

    @Override
    public <T> Optional<T> getSingle(SQLBuilder sqlBuilder, Connection connection, SQLBuilder.RowMapper<T> rowMapper) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return Optional.ofNullable(rs.next() ? rowMapper.map(rs) : null);
        }
    }

    @Override
    public <T> T getSingle(SQLBuilder sqlBuilder, Connection connection, SQLBuilder.RowMapper<T> rowMapper, T defaultValue) throws SQLException {
        try (PreparedStatement ps = build(sqlBuilder, connection); ResultSet rs = ps.executeQuery()) {
            return rs.next() ? rowMapper.map(rs) : defaultValue;
        }
    }

}

File: core/src/main/java/com/metricstream/jdbc/JdbcSQLBuilderProvider.kt
/*
 * Copyright © 2020-2022, MetricStream, Inc. All rights reserved.
 */
package com.metricstream.jdbc

import com.metricstream.jdbc.SQLBuilder.Companion.close
import com.metricstream.jdbc.SQLBuilder.Companion.wrapConnection
import com.metricstream.jdbc.SQLBuilder.Companion.wrapStatement
import kotlin.Throws
import java.sql.SQLException
import java.sql.PreparedStatement
import java.sql.ResultSet
import com.metricstream.jdbc.SQLBuilder.Masked
import java.math.BigDecimal
import java.sql.Connection
import java.sql.Date
import java.sql.Timestamp
import java.time.OffsetDateTime
import java.time.Instant
import java.util.Optional
import java.util.ServiceLoader

internal class JdbcSQLBuilderProvider : SQLBuilderProvider {

    @Throws(SQLException::class)
    private fun build(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        vararg columns: String
    ): PreparedStatement {
        val expanded: MutableList<Any?> = mutableListOf()
        sqlBuilder.interpolate(SQLBuilder.Mode.EXPAND_AND_APPLY, expanded)
        val ps: PreparedStatement = if (columns.isEmpty()) {
            connection.prepareStatement(
                sqlBuilder.statement.toString(),
                sqlBuilder.resultSetType,
                sqlBuilder.resultSetConcurrency
            )
        } else {
            connection.prepareStatement(
                sqlBuilder.statement.toString(),
                columns
            )
        }
        try {
            if (sqlBuilder.fetchSize > 0) {
                ps.fetchSize = sqlBuilder.fetchSize
            }
            if (sqlBuilder.maxRows >= 0) {
                ps.maxRows = sqlBuilder.maxRows
            }
            if (expanded.isNotEmpty()) {
                expanded.forEachIndexed { index, arg ->
                    when (arg) {
                        is LongString -> ps.setCharacterStream(index + 1, arg.reader)
                        is Masked -> ps.setObject(index + 1, arg.data)
                        else -> ps.setObject(index + 1, arg)
                    }
                }
            }
        } catch (ex: SQLException) {
            close(ps)
            throw ex
        }
        return ps
    }

    @Throws(SQLException::class)
    override fun getResultSet(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        wrapConnection: Boolean
    ): ResultSet {
        var preparedStatement: PreparedStatement? = null
        return try {
            preparedStatement = build(sqlBuilder, connection)
            preparedStatement.executeQuery()!!.let { rs ->
                if (wrapConnection) wrapConnection(rs) else wrapStatement(rs)
            }
        } catch (e: SQLException) {
            close(preparedStatement)
            throw e
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getInt(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Int
    ): Int {
        return get(
            sqlBuilder,
            connection,
            { it.getInt(columnNumber) },
            defaultValue
        )
    }

    private fun <T> get(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        transform: (ResultSet) -> T,
        defaultValue: T
    ): T {
        build(sqlBuilder, connection).use { ps ->
            ps.executeQuery().use { rs ->
                return if (rs.next()) transform(rs) else defaultValue
            }
        }
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getInt(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Int
    ): Int {
        return get(
            sqlBuilder,
            connection,
            { it.getInt(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getLong(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Long
    ): Long {
        return get(
            sqlBuilder,
            connection,
            { it.getLong(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getLong(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Long
    ): Long {
        return get(
            sqlBuilder,
            connection,
            { it.getLong(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDouble(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Double
    ): Double {
        return get(
            sqlBuilder,
            connection,
            { it.getDouble(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDouble(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Double
    ): Double {
        return get(
            sqlBuilder,
            connection,
            { it.getDouble(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getString(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: String?
    ): String? {
        return get(
            sqlBuilder,
            connection,
            { it.getString(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getBigDecimal(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: BigDecimal?
    ): BigDecimal? {
        return get(
            sqlBuilder,
            connection,
            { it.getBigDecimal(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getBigDecimal(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: BigDecimal?
    ): BigDecimal? {
        return get(
            sqlBuilder,
            connection,
            { it.getBigDecimal(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getString(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: String?
    ): String? {
        return get(
            sqlBuilder,
            connection,
            { it.getString(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getObject(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Any?
    ): Any? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getObject(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Any?
    ): Any? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDateTime(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: OffsetDateTime?
    ): OffsetDateTime? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnNumber, OffsetDateTime::class.java) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDateTime(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: OffsetDateTime?
    ): OffsetDateTime? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnName, OffsetDateTime::class.java) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getInstant(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Instant?
    ): Instant? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnNumber, OffsetDateTime::class.java).toInstant() },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getInstant(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Instant?
    ): Instant? {
        return get(
            sqlBuilder,
            connection,
            { it.getObject(columnName, OffsetDateTime::class.java).toInstant() },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getTimestamp(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Timestamp?
    ): Timestamp? {
        return get(
            sqlBuilder,
            connection,
            { it.getTimestamp(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getTimestamp(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Timestamp?
    ): Timestamp? {
        return get(
            sqlBuilder,
            connection,
            { it.getTimestamp(columnNumber) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnName The name of the column from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDate(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnName: String,
        defaultValue: Date?
    ): Date? {
        return get(
            sqlBuilder,
            connection,
            { it.getDate(columnName) },
            defaultValue
        )
    }

    /**
     * Returns a value from the first row returned when executing the query.
     * @param connection The Connection from which the PreparedStatement is created
     * @param columnNumber The index of the column (starting with 1) from which to return the value
     * @param defaultValue The default value that is returned if the query did not return any rows
     * @return the value from the query
     * @throws SQLException the exception thrown when generating or accessing the ResultSet
     */
    @Throws(SQLException::class)
    override fun getDate(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        columnNumber: Int,
        defaultValue: Date?
    ): Date? {
        return get(
            sqlBuilder,
            connection,
            { it.getDate(columnNumber) },
            defaultValue
        )
    }

    /**
     * Executes the SQL statement.
     * @param connection The Connection from which the PreparedStatement is created
     * @return The result of executeUpdate of that statement
     * @throws SQLException the exception thrown when executing the query
     */
    @Throws(SQLException::class)
    override fun execute(sqlBuilder: SQLBuilder, connection: Connection): Int {
        build(sqlBuilder, connection).use { ps -> return ps.executeUpdate() }
    }

    /**
     * Executes the SQL statement.
     * @param connection The Connection from which the PreparedStatement is created
     * @param keyColumns column names from the underlying table for which the inserted values will be returned.  Note that these names
     * not necessarily have to be part of the keyColumns into which the builder explicitly inserts values.
     * @return The result of getGeneratedKeys of that statement
     * @throws SQLException the exception thrown when executing the query
     */
    @Throws(SQLException::class)
    override fun execute(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        vararg keyColumns: String
    ): ResultSet {
        val ps = build(sqlBuilder, connection, *keyColumns)
        ps.executeUpdate()
        return wrapStatement(ps.generatedKeys)
    }

    @Throws(SQLException::class)
    override fun <T> getList(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        rowMapper: SQLBuilder.RowMapper<T>
    ): List<T> {
        build(sqlBuilder, connection).use { ps ->
            ps.executeQuery().use { rs ->
                return getList(rs, rowMapper, false)
            }
        }
    }

    @Throws(SQLException::class)
    override fun <T> getListWithNull(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        rowMapper: SQLBuilder.RowMapper<T?>,
    ): List<T?> {
        build(sqlBuilder, connection).use { ps ->
            ps.executeQuery().use { rs ->
                return getList(rs, rowMapper, true)
            }
        }
    }

    @Throws(SQLException::class)
    override fun <K, V> getMap(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        rowMapper: SQLBuilder.RowMapper<Map.Entry<K, V?>>,
        withNull: Boolean
    ): Map<K, V?> {
        build(sqlBuilder, connection).use { ps ->
            ps.executeQuery().use { rs ->
                return getMap(rs, rowMapper, withNull)
            }
        }
    }

    @Throws(SQLException::class)
    override fun <T : Any> getSingle(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        rowMapper: SQLBuilder.RowMapper<T?>
    ): Optional<T> {
        build(sqlBuilder, connection).use { ps ->
            ps.executeQuery().use { rs ->
                return Optional.ofNullable(if (rs.next()) rowMapper.map(rs) else null)
            }
        }
    }

    @Throws(SQLException::class)
    override fun <T> getSingle(
        sqlBuilder: SQLBuilder,
        connection: Connection,
        rowMapper: SQLBuilder.RowMapper<T?>,
        defaultValue: T?
    ): T? {
        return get(sqlBuilder, connection, { rowMapper.map(it) }, defaultValue)
    }

    // Lazy loading to avoid instantiating the connectionProviderImpl when a JdbcSQLBuilderProvider object is created.
    // This will change if and when a caller uses the connection-less access methods.
    override val connectionProvider: ConnectionProvider by lazy { connectionProviderImpl }

    companion object {
        // Defer searching for a provider implementation until the first usage of connectionProvider.
        private val connectionProviderImpl: ConnectionProvider by lazy {
            val service = ConnectionProvider::class.java
            ServiceLoader.load(service).firstOrNull() ?: throw IllegalStateException("Could not find an implementation of $service")
        }
    }
}
